### ISSUE TITLE
wstring (utf16) -> string (utf8) transition (community feedback requested)

### ISSUE THREAD

**Issue by axxel** [zxing-cpp/zxing-cpp#338]
As mentioned in e.g. #334, I intend to replace the usage of `std::wstring` in the `Result` struct with utf8 encoded `std::string` for better interoperability with non-windows systems. The question is: how to go about doing that with minimally annoying users of the library? I currently see these options:

1) introduce a preprocessor macro `ZX_USE_UTF8` that client code can define before including `ReadBarcode.h` that would switch the return type of `Result::text()` to `std::string`. The absence of this macro could trigger a deprecation compiler warning. In a second phase I could then default to using utf8 but introduce a `ZX_USE_UTF16` macro that switches back to `std::wstring` to allow a 1-line fix or even a build environment setting to go back to the old ways.

2) introduce 2 new `Result` properties: `std::string utf8()` and `std::wstring utf16()`, the latter being an alias for the current `text()`. This would not introduce a breaking change and lend itself to keep both variants indefinitively. If feels like less hassle but also a bit unusual and over the top. It is also no solution for the `std::wstring ecLevel()`, which will simply be an ASCII `std::string`. :-/

This question is also related to #333.

Any feedback on preferences or even other alternatives welcome.

**Comment by axxel** [comment]
It seems either no-one has read this post or no-one cares. Would anyone object if I skipped all the transition effort and simply changed the return type of `text`, thereby breaking all of your code? :-P

**Comment by axxel** [comment]
Thanks for your comment. I think `content` is not a good name as it does not convey a distinction to `bytes`. Both are representation of the content of the symbol. In fact I think the meaning 'human readable interpretation' is exactly what would be most helpful for most of the users. It is also at the core of the whole binary content discussion. And with that in mind I was already looking into bringing the GS1 "(ai)data" format from DataBar to generally all GS1 codes and ideally remove that internal transformation from the `bytes` channel where it certainly doesn't belong - except if that was actually specified in the DataBar spec, which I did not check, yet, but seemed unlikely.

As an alternative to simply doing that translation to all GS1 codes, one could argue to introduce a `DecodeHints` flag to enable it or to supply a separate helper function that can be used to make that transformation.

**Comment by vkrause** [comment]
> Any feedback on preferences or even other alternatives welcome.

From the KDE consumer point-of-view where we have to support multiple ZXing versions provided by Linux distributors for some time usually: The second option seems a bit easier and slightly more predictable, but since we have to wrap the usage in version check ifdefs for some time anyway the difference isn't all that big either.

**Comment by gitlost** [comment]
Re GS1 it is defined that way in the GS1 General Specifications, i.e. the data is the AIs followed by their data fields, with `<GS>` separators inserted where required. It is not the "(ai)data" form, which is the Human Readable Interpretation, which is intended to be printed with the barcode in case manual entry is required. (There is another form, what they unhelpfully call "Non-HRI Text", which is an expanded version of HRT with the AIs labeled - see GS1 General Specifications 22.0 (2022), Section 4.15.)

On the general point, I haven't followed the text/binary discussion closely but my general impression is that it's a wrong-headed distinction. There's the byte stream - what is now named `bytes()` (good name). There is the bytes interpreted according to the character encodings (specified or otherwise) - what is now named `text()` (bad name). There should also be a separate API returning the GS1 HRI, which would return the "(ai)data" format, named `hri()` or something. It would also be nice to supply an API returning the GS1 "Non-HRI Text" form.

**Comment by gitlost** [comment]
Re HRI form it could also deal with e.g. ISBN formats (with dashes inserted), and other non-GS1 formatting.

**Comment by axxel** [comment]
It seems non-controversial that `bytes` or actually `bytesECI` is or rather should be the basic ISO conform 'content' of the symbol. (By the way: I'm ok with the term `bytesECI` but am open for creative alternatives). Then we have conceptually two more views:
 * utf8 version (of `bytes`)
 * HRI version (of utf8 version of `bytes`)

I'm questioning how pronounced these two variants need to be represented in the API?
 1. just provide the HRI version
 2. provide a hint to switch between the two
 3. provide two separate 'channels'

The choice of name(s) is dependent on which of these end up being used. Since we currently have 1) and no-one was asking for more, I did not feel the need to provide even more channels than the 4 we have right now. But since I don't want to break the API anytime soon again, I'd rather end up with something lasting, i.e. capable enough.

I'd like to hear from people actually using this in production code/apps what their use cases are with respect to the different 'types' of content. @markusfisch @vkrause @huycn (if I should stop mentioning individuals in these kind of potentially annoying "say something" pokes, please let me know ;)).

**Comment by axxel** [comment]
@vkrause thanks. What is your stance on the `utf8` vs. `HRI` question? How do those two fit into your landscape? Do have a use for both, if not, which one then?

**Comment by axxel** [comment]
Update on my current plan regarding the successor of `text`: introduce `std::string utf8()` and `std::string hri()` (the full name `humanReadableInterpretation()` is just too ugly for my taste).

That leaves the question how to deal with the remaining `std::wstring ecLevel()`?
 1. introduce `eccLevel`
 2. introduce `errorCorrectionLevel`
 3. simply replace the return type, introducing an API braking change without deprecation warning
 4. do the whole macro dance describe above (I don't see that coming)

A google search shows actually 4x the results for `eccLevel` compared to `ecLevel` but the QRCode spec uses the term "error correction level" not "error correction capacity level".

I assume very few people are making use of this member anyway so maybe option 3 is practical. As usual: comments welcome.

**Comment by oehhar** [comment]
Am 18.06.2022 um 16:32 schrieb axxel:
> Update on my current plan regarding the successor of |text|: introduce 
> |std::string utf8()| and |std::string hri()| (the full name 
> |humanReadableInterpretation()| is just too ugly for my taste).
> 
> That leaves the question how to deal with the remaining |std::wstring 
> ecLevel()|?
> 
>  1. introduce |eccLevel|
>  2. introduce |errorCorrectionLevel|
>  3. simply replace the return type, introducing an API braking change
>     without deprecation warning
>  4. do the whole macro dance describe above (I don't see that coming)
> 
> A google search shows actually 4x the results for |eccLevel| compared to 
> |ecLevel| but the QRCode spec uses the term "error correction level" not 
> "error correction capacity level".

My personal perception is, that "ECC" means "Error correction Code" and 
is a well established term. But I am in Data Matrix, not in QR.

Maybe, the QR code spec may be enhanced here.

THanks,
Harald

**Comment by axxel** [comment]
> Maybe, the QR code spec may be enhanced here.

I don't see any issue with the terms used there. Both QR and DM spec use the term "error correction level" (so neither "error correction capacity level" nor "error correction code level") hence, I would not have chosen `eccLevel` when starting from scratch but `ecLevel` (or the long form). This led me to double check my google search and it turns out, I failed at using google!?! I also gave the github search a go and both agree that `ecLevel` is by far the more popular term (no wonder, as it is more accurate).

So this leaves option 2 and 3 on the table. I vote for 3 (just switch the return type). Those who agree, please leave a thumbs up, the others a thumbs down.

**Comment by axxel** [comment]
Sorry people but I have to admit that I am turning in circles here. When making up my mind about the name `utf8` I was too narrowly focused on the c++ `Result` class and the distinction between `bytes` and the "utf8 version" thereof and since it seemed an easy way out of the deprecation issues, I liked the idea (better than any other I had/heard that far). Now properly taking the wrappers into account, it is obvious to me that `utf8` does not make any sens there at all. We still need a name for that as well and if that stays `text` (which it could without disruption as there is no type transition required) then there is a strong argument to use it on the c++ side as well, if only for consistency.

The only other meaningful alternative I could come up with is `str` or `string` (see e.g. the python type). I agree with @gitlost that the term `text` semi-implies 'human readable', which this blob of data will not strictly be. It feels like `string` carries that meaning less so but therefore focuses more on the 'data type' instead of the 'content type' which is unfortunate. The problem would be solved if we did not need this channel at all and only provided `bytes` and `text` where the latter would mean the HRI. I have the gut-feeling that those 2 variants are the most helpful anyway: Those who need the 'raw' or the 'standard' data will go for `bytes` and those who end up displaying a string on screen will be better served with the HRI (maybe with some escaping of non-graphical characters like the `-escape` option of `ZXingReader` for anything non-text).

I could still let there be the current `utf8` property but it will serve a yet to be found use case and not be exposed in any of the wrappers (until demand is expressed).

Notes:
 * the term `utf8ECI` does not have this problem as it does explicitly contain exactly that (utf8 encoded bytes according to the ECI protocol)
 * `text` is an established name and is also used by at least one commercial SDK
 * if it was not for the type transition issue, I would not have considered changing it

I read the relative silence from you guys on this matter as a sign that it is just not that important to you how this one function is called, which I understand completely...

**Comment by axxel** [comment]
To make some progress here, I decided to go the original `#define ZX_USE_UTF8` route and keep `text` and only switch the return type. To make that an ABI compatible, client side switch, `text` is inline and either calls `utf8` or `utf16`. Whether those two stay after the completed transition will be decided later. Same goes for the question of whether `text` will return a HRI version of `utf8` or if a new `hri` will be added for that.

This should be close to release 1.4 now to get the deprecation warnings out and about. Please merge and test your code and let me know if something broke. Thanks.

Before releasing, I'll make sure the android, python and Qt wrappers will have access to the `contentType` and both `bytes` and `text`.

**Comment by axxel** [comment]
@gitlost @oehhar I'm about to push a new API that translates `GS1` strings to their HRI version (extracting it from the DataBar code base). I'm not happy with the naming ideas I came up so far. The main issue being that I don't know how these strings are called within the `GS1 System`. I glanced at the spec but could not find a proper term. My current ideas:

 * `HRIFromGS1`
 * `HRIFromGS1AIs`
 * `HRIFromGS1System`
 * `HRIFromGS1String`

Any better ideas?

**Comment by oehhar** [comment]
I prefer HRIFromGS1. No better idea.
Harald

**Comment by axxel** [comment]
Thanks for your quick feedback. It is a freestanding global function that is meant to be later used in `Result::text()` or `Result::hri()`. See 24b737365358bdb0e404e65cb70d0071362159f7.

**Comment by axxel** [comment]
@gitlost Reimplementing the DataBarExpanded bit stream decoder to get rid of both `DecodeStatus` usage and the hard-coded HRI stuff going on there, I came across two issues:

 1. `rssexpanded-3/13.png` contains the 'raw' substring `3922795888888888888888888888888888888888888888888888888888`. 3922 has a max length of 15 according to the spec. My new `HRIFromGS1` fails then decode this. My interpretation: this is an invalid symbol.
 2.  There are a bunch of samples that contain the 422 field which is fixed size 3 but then have a FNC1 follow it. Again invalid according to my understanding.
 
Comments?

**Comment by gitlost** [comment]
Re 1. yes this is invalid GS1. How to treat this? - not sure.

Re 2. there's only a set number of AIs which are considered "predefined length", the first 2 digits of which are listed in GS1 General Specifications v22.0 Figure 7.8.5-2, and "42" is not one of them, so "422" should be terminated by an FNC1, even though it has a fixed length. (I think the idea is that the "predefined length" AIs will never be added to the future, i.e. are set in stone.)

Note also Section 7.8.6.3: "Notwithstanding the above, the processing routine SHALL tolerate a single separator character
immediately following any element string, whether necessary or not..."

(The handiest and best overview I think of AIs is given by BWIPP's [gs1-format-spec.txt](https://github.com/bwipp/postscriptbarcode/blob/master/contrib/development/gs1-format-spec.txt) - generating the `aiInfos` table in "GS1.cpp" automatically from this is on my to-do list (eye-balled it the last time)).

**Comment by axxel** [comment]
Thanks a lot. Very help feedback. I fixed 2 and `HRIFromGS1` does indeed currently return an empty string. This means a couple blackbox tests currently fail until I change them to test for proper `bytes()` content instead of `HRIFromGS1()`. Will do that before release 2.0. Which will be a breaking change for some. Nothing to be done about that, I'm afraid....

**Comment by axxel** [comment]
To be a bit more flexible for the future and allow for even more 'rendering' options without introducing ever more 'channels' I'd like to mention the possibility to parameterize `text( TextMode mode = TextMode::HRI )`. With modes like
 * `Utf8` - simply return what `utf8()` currently does
 * `HRI` - human readable interpretation
 * `Hex` - `ToHex(bytes)`
 * `Base64` - maybe?
 * `EscapeNonPrintable` - what `ZXingReader -escape` currently does
 * `Utf8ECI` - replace `utf8ECI()`?

This does not translate too well into the wrappers though as it is way easier to return a POD struct with some ready made strings than to pass an object that can sill call into c++ code. So there we'd have the 'channels' either way, I think. Hm, we could also have both, of course...

I'll implement this in the backend (`Content`) where it makes sens either way and also see if there is any feedback from others about this idea and then decide.

Looks like my plan to release 1.4 today has to be postponed again until I made up my mind about this.

**Comment by axxel** [comment]
I decided to postpone the decision but get 1.4 out anyway. I realized that there is another related issue and that is the fallback character set that can currently be specified via `DecodeHints::setCharacterSet`. That has the same property as the outlined `TextMode` in the sense that it actually does not really belong in the `DecodeHints` because it is only needed during the `bytes` -> `text` transition and could therefore actually be a parameter of `text()` as well. In fact it could very well be that the user might want to specify a different character set for different symbols/results from the same `ReadBarcodes()` call (especially if they are of different symbologies). That might be very rare, though and it is convenient from the wrappers perspective to handle it like it is handled now (even though that is more for historic reasons).

Bottom line: if specifying the character set via `DecodeHints` stays, then specifying a `TextMode` via the same route would be consistent and pragmatic for the wrappers. While at the same time we could provide a more capable `Result::text(TextMode mode = HRI, std::string fallbackCharset = {})` function that would then override anything provided via `DecodeHints` beforehand.

All not very 'clean' I admit. Better ideas someone?

**Comment by axxel** [comment]
The polls are in: no better ideas have been presented. ;). I went ahead and tuned the constants a little:
```
enum class TextMode : unsigned char // see above
{
	Raw,     ///< bytes() transcoded to unicode/utf8 based on ECI info or guessed based on content
	ECI,     ///< standard content following the ECI protocol with every character set ECI segment transcoded to unicode/utf8
	HRI,     ///< Human Readable Interpretation (dependent on the ContentType)
	Hex,     ///< bytes() transcoded to ASCII string of HEX values
	Escaped, ///< Use the TextUtfEncoding::EscapeNonGraphical() function (e.g. ASCII 29 will be transcoded to "<GS>")
};
```
I decided to remove the mentioning of 'utf8' as that does not make sense for the wrappers where strings may be encoded in another way depending on the platform. What I'm still struggling somewhat are two questions:
 1. Better term for `Raw`? Maybe `Plain` or `Unicode`? (Decided against `Default` and `Standard`)
 2. What should the default setting for this be? I'd argue `HRI`, because users who don't know much about the subject likely just want to display a string somewhere and will not find/look for the `textMode` setting. But this would be a potentially surprising/breaking change for a couple of users. It would also require changes to the test, as a whole bunch would then fail.

**Comment by axxel** [comment]
I went for the term `Plain` and used it as the default (for now). feedback welcome.

**Comment by axxel** [comment]
Let's see if the default mode change to `HRI` will 'provoke' some feedback ;).

**Comment by markusfisch** [comment]
Sorry if I have delayed closing this 🙈 And sorry for not giving feedback earlier. I didn't really understand the scope of this change.

**Comment by axxel** [comment]
@markusfisch You did not delay anything here. Your recent commit merely reminded me that this was still open and that now might be a reasonable time to call it done.
